### DISCUSSION TITLE
broadcast presentation mode

### DISCUSSION THREAD

**Discussion by chrisns** [marp-team/marp#273] [Ideas]
Happy to work on it, just floating the idea before I sink the time in it rather than making a local hack:

Thinking it could be great to use webRTC with STUN/TURN to allow others watching my presentation to do so in a browser and have their view sync'd to the presenters, also handy when doing a talk somewhere that has an av setup who just want a 'copy of my slides' which they load on to some back office whatever, I could just give them an address and still drive from my computer (if I publish my slides somewhere public or on a lan), or just a poor-man's hdmi splitter :)

similar to how [spectacle](https://formidable.com/open-source/spectacle/docs/basic-concepts/) does with `?presenterMode=true`

thoughts?

**Comment by yhatt** [comment]
As a premise, HTML output is a bonus for us, not a core feature. Marp team want to focus on the experience of slide creation, and leave features to support presentation to the other tool as much as possible.

Nevertheless, I'm interested in sharing the slide state over (P2P) network. It not only provides a way of broadcast but also may turn a mobile device into a remote for the slide.

I have thought about WebRTC powered P2P [state sync](https://github.com/marp-team/marp-cli/blob/main/src/templates/bespoke/sync.ts) over once or twice before, but I could not find out a compromise of implementation. Followings are (not organized) memo:

> - Many Marp users like to present slides quickly from the local file, and they may feel reluctant extra actions to run HTTPS server. So working WebRTC from the local file would be best (e.g. HTML file sent by email like a ppt file).
>   - WebRTC requires HTTPS or localhost so it looks like a dream. But refering to the remote WebRTC controller `<iframe src="https://...." />` and communicating with `postMessage()` may make the dream be true.
> - To keep maintaining both of signaling server and TURN is hard for us now!
>   - Signaling server may not be needed if got some idea to pass the connection data to subscribers. (URL query?)
>   - I want to know some of SaaS for WebRTC :)
> - :arrow_right: Currently Marp team would not pay off working for these...

Continue to discuss :point_down:

**Reply by chrisns** [reply]
theres public free stun servers knocking about, and some turn https://gist.github.com/sagivo/3a4b2f2c7ac6e1b5267c2f1f59ac6c6b

maybe the key is a plugin arch for the cli to be able to add arbitrary js+css rather than have to roll it into the main marp repos, allowing things like this to come from the community without taking away from your focus?

and yeah, mobile device as remotes, and/or using a tablet with the notes resized to fill the page😉 for a teleprompter

could probably make a smartwatch something something to navigate slides etc
